clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java

```diff
@@ -0,0 +1,149 @@
+                                                               CloudFoundryClusterProvider clusterProvider) {
+    this.operationPoller = operationPoller;
+    this.credentialsRepository = credentialsRepository;
+    this.clusterProvider = clusterProvider;
+  }
+
+  @Override
+  public AtomicOperation convertOperation(Map input) {
+    return new DeployCloudFoundryServerGroupAtomicOperation(operationPoller, convertDescription(input), clusterProvider);
+  }
+
+  @Override
+  public DeployCloudFoundryServerGroupDescription convertDescription(Map input) {
+    DeployCloudFoundryServerGroupDescription converted = getObjectMapper().convertValue(input, DeployCloudFoundryServerGroupDescription.class);
+    CloudFoundryCredentials credentials = getCredentialsObject(input.get("credentials").toString());
+    converted.setClient(credentials.getClient());
+    converted.setAccountName(credentials.getName());
+
+    converted.setSpace(findSpace(converted.getRegion(), converted.getClient())
+      .orElseThrow(() -> new IllegalArgumentException("Unable to find space '" + converted.getRegion() + "'.")));
+
+    Map artifactSource = (Map) input.get("artifactSource");
+
+    if ("artifact".equals(artifactSource.get("type"))) {
+      ArtifactCredentials artifactCredentials = credentialsRepository.getAllCredentials().stream()
+        .filter(creds -> creds.getName().equals(artifactSource.get("account")))
+        .findAny()
+        .orElseThrow(() -> new IllegalArgumentException("Unable to find artifact credentials '" + artifactSource.get("account") + "'"));
+
+      converted.setArtifact(convertToArtifact(artifactSource.get("account").toString(), artifactSource.get("reference").toString()));
+      converted.setArtifactCredentials(artifactCredentials);
+    } else if ("package".equals(artifactSource.get("type"))) {
+      CloudFoundryCredentials accountCredentials = getCredentialsObject(artifactSource.get("account").toString());
+      converted.setArtifactCredentials(new PackageArtifactCredentials(credentials.getClient()));
+
+      Artifact artifact = new Artifact();
+      artifact.setType("package");
+      artifact.setReference(getServerGroupId(artifactSource.get("serverGroupName").toString(),
+        artifactSource.get("region").toString(), accountCredentials.getClient()));
+      converted.setArtifact(artifact);
+    }
+
+    Map manifest = (Map) input.get("manifest");
+    if ("direct".equals(manifest.get("type"))) {
+      converted.setApplicationAttributes(getObjectMapper().convertValue(manifest,
+        DeployCloudFoundryServerGroupDescription.ApplicationAttributes.class));
+    } else if ("artifact".equals(manifest.get("type"))) {
+      Artifact manifestArtifact = convertToArtifact(manifest.get("account").toString(), manifest.get("reference").toString());
+      ArtifactCredentials manifestArtifactCredentials = credentialsRepository.getAllCredentials().stream()
```

Comment:
I think this logic should live in the artifacts module in clouddriver -- finding the credentials & downloading artifacts is a common-enough paradigm to share the code.

Reply:
This is being replaced by the expected artifacts improvement in the end.